Content/Items/Misc/Weapons.Sling.cs

```diff
@@ -0,0 +1,491 @@
+            }
+            if (Projectile.velocity.Y > 16f)
+                Projectile.velocity.Y = 16f;
+        }
+
+        public override bool PreDraw(ref Color lightColor)
+        {
+            Main.instance.LoadItem(ItemID.Mushroom);
+            Texture2D tex = TextureAssets.Item[ItemID.Mushroom].Value;
+            Vector2 drawOrigin = tex.Size() / 2f;
+            for (int k = 0; k < Projectile.oldPos.Length; k++)
+            {
+                Vector2 drawPos = (Projectile.oldPos[k] - Main.screenPosition) + drawOrigin + new Vector2(0f, Projectile.gfxOffY);
+                Color color = Projectile.GetAlpha(lightColor) * ((Projectile.oldPos.Length - k) / (float)Projectile.oldPos.Length);
+                Main.EntitySpriteDraw(tex, drawPos, null, color, Projectile.rotation, drawOrigin, Projectile.scale, SpriteEffects.None, 0);
+            }
+            return true;
+        }
+
+        public override void OnHitNPC(NPC target, int damage, float knockback, bool crit)
+        {
+            target.AddBuff(BuffID.Poisoned, Main.rand.Next(new int[] { 180, 240, 300 }));
+        }
+
+        public override void Kill(int timeLeft)
+        {
+            for (int i = 0; i < 20; i++)
+            {
+                Dust.NewDustDirect(Projectile.position, Projectile.width, Projectile.height, ModContent.DustType<Dusts.MushroomDust>()).scale = Main.rand.NextFloat(0.8f, 1.2f);
+            }
+
+            Terraria.Audio.SoundEngine.PlaySound(SoundID.Grass with { Pitch = -0.1f, Volume = 1.25f }, Projectile.Center);
+        }
+    }
+
+    class SlingStoneProjectile : ModProjectile
+    {
+        public override string Texture => "Terraria/Images/Item_" + ItemID.StoneBlock;
+
+        public override void SetStaticDefaults()
+        {
+            DisplayName.SetDefault("Stone");
+            ProjectileID.Sets.TrailingMode[Projectile.type] = 0;
+            ProjectileID.Sets.TrailCacheLength[Projectile.type] = 7;
+        }
+        public override void SetDefaults()
+        {
+            Projectile.friendly = true;
+            Projectile.DamageType = DamageClass.Ranged;
+
+            Projectile.timeLeft = 600;
+            Projectile.penetrate = 1;
+
+            Projectile.width = Projectile.height = 16;
+        }
+        public override void AI()
+        {
+            Projectile.rotation += (0.35f * (Projectile.velocity.X * 0.15f)) * Projectile.direction;
+            Projectile.velocity.Y += 0.2f;
+            if (Projectile.velocity.Y > 0)
+            {
+                if (Projectile.velocity.Y < 13f)
+                    Projectile.velocity.Y *= 1.085f;
+                else
+                    Projectile.velocity.Y *= 1.04f;
+            }
+            if (Projectile.velocity.Y > 16f)
+                Projectile.velocity.Y = 16f;
+        }
+
+        public override bool PreDraw(ref Color lightColor)
+        {
+            Main.instance.LoadItem(ItemID.StoneBlock);
```

Comment:
This seems dangerous to put in draw. Potentially grab texture at runtime directly, and cache it